nx/lib/nx/random.ex

```diff
@@ -601,6 +601,242 @@ defmodule Nx.Random do
+      iex> result
+      #Nx.Tensor<
+        s64[4][5]
+        [
+          [1, 1, 1, 1, 0],
+          [4, 4, 4, 4, 3],
+          [7, 7, 7, 7, 6],
+          [10, 10, 10, 10, 9]
+        ]
+      >
+      iex> {result, _key} = Nx.Random.choice(k, t, p, samples: 3, axis: 1, replace: false) # without replacement
+      iex> result
+      #Nx.Tensor<
+        s64[4][3]
+        [
+          [1, 2, 0],
+          [4, 5, 3],
+          [7, 8, 6],
+          [10, 11, 9]
+        ]
+      >
+
+  If no axis is specified, the tensor is flattened.
+  Notice that in the first case we get a higher occurence
+  of the entries with bigger probabilities, while in the
+  second case, without replacements, we get those samples
+  first.
+
+      iex> k = Nx.Random.key(2)
+      iex> t = Nx.iota({2, 3})
+      iex> p = Nx.tensor([0.01, 0.1, 0.19, 0.6, 0.05, 0.05])
+      iex> {result, _key} = Nx.Random.choice(k, t, p, samples: 10) # with replacement
+      iex> result
+      #Nx.Tensor<
+        s64[10]
+        [2, 1, 3, 3, 3, 1, 3, 3, 1, 2]
+      >
+      iex> {result, _key} = Nx.Random.choice(k, t, p, samples: 6, replace: false) # without replacement
+      iex> result
+      #Nx.Tensor<
+        s64[6]
+        [3, 1, 2, 5, 4, 0]
+      >
+  """
+  deftransform choice(key, tensor, p), do: choice_probabilities(key, tensor, p)
+  deftransform choice(key, tensor, p, opts), do: choice_probabilities(key, tensor, p, opts)
+
+  deftransformp validate_choice_opts(tensor, opts) do
+    opts = Keyword.validate!(opts, [:samples, :axis, replace: true])
+
+    {axis, tensor_shape} =
+      case opts[:axis] do
+        nil ->
+          {0, {Tuple.product(tensor.shape)}}
+
+        axis ->
+          {Nx.Shape.normalize_axis(tensor.shape, axis, tensor.names), tensor.shape}
+      end
+
+    if Nx.rank(tensor) < 1 do
+      raise ArgumentError, "tensor must have rank 1 or greater"
+    end
+
+    n_draws = opts[:samples]
+
+    if n_draws < 1 do
+      raise "must take at least one sample, got samples=#{n_draws}"
+    end
+
+    n_inputs =
+      case opts[:axis] do
+        nil -> Nx.size(tensor)
+        _ -> Nx.axis_size(tensor, axis)
+      end
+
+    replace = opts[:replace]
+
+    if not replace and n_draws > n_inputs do
+      raise ArgumentError, "cannot take more samples than the input size when replace: false"
+    end
+
+    {tensor_shape, n_inputs, n_draws, axis, replace}
+  end
+
+  defnp choice_uniform(key, tensor, opts) do
+    {tensor_shape, n_inputs, n_draws, axis, replace} = validate_choice_opts(tensor, opts)
+    tensor = Nx.reshape(tensor, tensor_shape)
+
+    if replace do
+      {idx, key} = randint(key, 0, n_inputs, shape: {n_draws})
+      result = Nx.take(tensor, idx, axis: axis)
+      {result, key}
+    else
+      {shuffled, key} = shuffle(key, tensor, axis: axis)
+      result = Nx.slice_along_axis(shuffled, 0, n_draws, axis: axis)
+      {result, key}
+    end
+  end
+
+  defnp choice_probabilities(key, tensor, p, opts \\ []) do
```

Comment:
It seems options are always required, so I think you should do this:

```suggestion
  defn choice(key, tensor, p, opts) do
```

Reply:
indeed they are :)
probabilities are not, so I'll still keep `choice_probabilities/choice_uniform` as separate `defnp`s

Reply:
No need for this actually :)